server/src/main/java/org/apache/druid/catalog/model/table/BaseInputSourceDefn.java

```diff
@@ -0,0 +1,278 @@
+
+  /**
+   * Overridden by each subclass to return the input source class to be
+   * used for JSON conversions.
+   */
+  protected abstract Class<? extends InputSource> inputSourceClass();
+
+  @Override
+  public void bind(TableDefnRegistry registry)
+  {
+    this.adHocTableFn = defineAdHocTableFunction();
+  }
+
+  @Override
+  public void validate(ResolvedExternalTable table)
+  {
+    convertTableToSource(table);
+  }
+
+  /**
+   * Overridden by each subclass to define the parameters needed by each
+   * input source.
+   */
+  protected abstract AdHocTableFunction defineAdHocTableFunction();
+
+  @Override
+  public TableFunction adHocTableFn()
+  {
+    return adHocTableFn;
+  }
+
+  /**
+   * Define a table "from scratch" using SQL function arguments.
+   */
+  protected ExternalTableSpec convertArgsToTable(
+      final Map<String, Object> args,
+      final List<ColumnSpec> columns,
+      final ObjectMapper jsonMapper
+  )
+  {
+    return new ExternalTableSpec(
+        convertArgsToSource(args, jsonMapper),
+        convertArgsToFormat(args, columns, jsonMapper),
+        Columns.convertSignature(columns)
+    );
+  }
+
+  /**
+   * Convert the input source using arguments to a "from scratch" table function.
+   */
+  protected InputSource convertArgsToSource(Map<String, Object> args, ObjectMapper jsonMapper)
+  {
+    final Map<String, Object> jsonMap = new HashMap<>();
+    auditInputSource(jsonMap);
+    convertArgsToSourceMap(jsonMap, args);
+    return convertSource(jsonMap, jsonMapper);
+  }
+
+  /**
+   * Convert SQL arguments to the corresponding "generic JSON" form in the given map.
+   * The map will then be adjusted and converted to the actual input source.
+   */
+  protected abstract void convertArgsToSourceMap(Map<String, Object> jsonMap, Map<String, Object> args);
+
+  /**
+   * Convert SQL arguments, and the column schema, to an input format, if required.
+   */
+  protected InputFormat convertArgsToFormat(Map<String, Object> args, List<ColumnSpec> columns, ObjectMapper jsonMapper)
+  {
+    return null;
+  }
+
+  /**
+   * Complete a partial table using the table function arguments and columns provided.
+   * The arguments match the set of parameters used for the function. The columns are
+   * provided if the SQL included an {@code EXTENDS} clause: the implementation should decide
+   * if columns are required (or allowed) depending on whether the partial spec already
+   * defines columns.
+   *
+   * @param table   the partial table spec, with input source and format parsed into a
+   *                generic Java map
+   * @param args    the argument values provided in the SQL table function call. The arguments
+   *                use the Java types defined in the parameter definitions.
+   * @param columns the set of columns (if any) from the SQL {@code EXTEND} clause
+   *
+   * @return an external table spec which Calcite can consume
+   */
+  protected abstract ExternalTableSpec convertCompletedTable(
```

Comment:
This this be moved into the `PartialTableFunction` class, since it seems it should only be used in that case?